cpp/src/arrow/json/parser.cc

```diff
@@ -0,0 +1,1209 @@
+    default:
+      return Status::NotImplemented("Unknown Builder");
+  }
+}
+
+class TypedHandler
+    : public rapidjson::BaseReaderHandler<rapidjson::UTF8<>, TypedHandler> {
+ public:
+  TypedHandler(StructBuilder* root_builder) : builder_(root_builder) {}
+
+  bool Null() {
+    if (Skipping()) return true;
+    // TODO check whether we're currently nullable
+    status_ = VisitBuilder(builder_, AppendNullVisitor{});
+    return status_.ok();
+  }
+
+  bool Bool(bool value) {
+    if (Skipping()) return true;
+    if (builder_->type()->id() == Type::BOOL) {
+      status_ = static_cast<BooleanBuilder*>(builder_)->Append(value);
+    } else {
+      status_ = ConversionError();
+    }
+    return status_.ok();
+  }
+
+  bool RawNumber(const char* data, rapidjson::SizeType size, bool) {
+    if (Skipping()) return true;
+    status_ = VisitBuilder(builder_, AppendNumberVisitor{data, size});
+    return status_.ok();
+  }
+
+  bool String(const char* data, rapidjson::SizeType size, bool) {
+    if (Skipping()) return true;
+    status_ = VisitBuilder(builder_, AppendStringVisitor{data, size});
+    return status_.ok();
+  }
+
+  bool StartObject() {
+    ++depth_;
+    if (Skipping()) return true;
+    if (builder_->type()->id() == Type::STRUCT) {
+      auto struct_builder = static_cast<StructBuilder*>(builder_);
+      auto num_fields = struct_builder->num_fields();
+      status_ = struct_builder->Append();
+      builder_stack_.push_back(builder_);
+      absent_fields_.push_back(std::vector<bool>(num_fields, true));
+    } else {
+      status_ = ConversionError();
+    }
+    return status_.ok();
+  }
+
+  bool Key(const char* key, rapidjson::SizeType len, bool) {
+    MaybeStopSkipping();  // new key at the depth where we started skipping ->
+                          // terminate skipping
+    if (Skipping()) return true;
+    auto parent = static_cast<StructBuilder*>(builder_stack_.back());
+    auto parent_type = std::static_pointer_cast<StructType>(parent->type());
+    auto field_index = parent_type->GetChildIndex(std::string(key, len));
+    if (field_index == -1) {
+      skip_depth_ = depth_;
+      return true;
+    }
+    absent_fields_.back()[field_index] = false;
+    builder_ = parent->field_builder(field_index);
+    return true;
+  }
+
+  bool EndObject(rapidjson::SizeType) {
+    MaybeStopSkipping();  // end of object containing depth where we started
+                          // skipping -> terminate skipping
+    --depth_;
+    if (Skipping()) return true;
+    int field_index = 0;
+    auto parent = static_cast<StructBuilder*>(builder_stack_.back());
+    for (bool null : absent_fields_.back()) {
+      // TODO since this is expected to be sparse, it would probably be more
+      // efficient to use CountLeadingZeros() to find the indices of the few
+      // null fields
+      if (null) {
+        status_ = VisitBuilder(parent->field_builder(field_index), AppendNullVisitor{});
```

Comment:
I'd say for performance let's avoid virtual, but it would definitely be more user friendly to have a "works-for-all-builders" set of append functions. Those could use VisitBuilderInline under the hood, and maybe they should be non-members to avoid obfuscating general [requirement that builders be downcast before use](https://github.com/apache/arrow/blob/master/cpp/src/arrow/array/builder_base.h#L59-L60):

```c++
/// \brief Downcast builder as necessary and append null
inline Status AppendNull(ArrayBuilder* builder) {
  struct {
    // Visit(...)
  } visitor;
  return VisitBuilderInline(builder, &visitor);
}
```

(see also [ARROW-4067 RFC: standardize ArrayBuilder subclasses](https://issues.apache.org/jira/browse/ARROW-4067))